### PR TITLE
Orion code updates

### PR DESCRIPTION
**Pull request by jamesdreid**
Made numerous updates to the Stackstorm Orion pack, including:

Corrected numerous typos and inaccurate descriptions in the code
Corrected logic in the main action.py lib to allow integration with Solarwinds installs that do not include NCM
Updated requirements to use current version of orionsdk
Added new Actions
- Add SNMPv3 Node
- Discover Interfaces by Type and Name with Admin Status Up requirement option
- Update Interface properties
- Update Interface custom properties
- Corrected logic in Node Poll Now action so it functions properly and includes data for sites in a "Warning (3)" state

### CODE REVIEW

File: pack.yaml
Comment (nmaludy) [inline pack.yaml]:
```suggestion
version: 0.8.0
```

File: requirements.txt
Comment (nmaludy) [inline requirements.txt]: We should continue to pin the orionsdk version, it's fine to update it, but removing the pin altogether is asking for trouble.

Comment (jjm) [inline requirements.txt]: Instead of removing the pin, it would be better to change it so that it'll auto update 0.x.y and not pull in 1.0.0.

```suggestion
orionsdk~=0.3
```

File: pack.yaml
Comment (jjm) [inline pack.yaml]: Is the next version not 0.8.0?

```suggestion
version: 0.8.0
```

File: requirements.txt
Comment (jamesdreid) [reply requirements.txt]: Makes sense.  Suggested limit added to the requirement.txt file

File: pack.yaml
Comment (jamesdreid) [reply pack.yaml]: I still have lots to learn about versioning.  I had bumped it to 0.8.0 initially but then I went back and added 2 more actions to my repo so I bumped the Rev from 0 to 4 to reflect those changes.  I have dropped it back to 0.8.0

**Comment by jamesdreid** [comment]
Any update on this request?  Is there anything else that I need to resolve to get it merged?

File: actions/lib/actions.py
Comment (jjm) [inline actions/lib/actions.py]: This exception is too board, it should really only catch the expected error.

File: actions/lib/actions.py
Comment (jjm) [inline actions/lib/actions.py]: This would be cleaner as a single line:

```suggestion
                self.logger.info("Connection to NCM failed. NCM not installed")
```

File: actions/node_create_snmpv3.py
Comment (jjm) [inline actions/node_create_snmpv3.py]: I likely wrote the above code in `NodeCreate()`, the something like this would be cleaner and easier to understand.

```suggestion
        engineID = 1
        if poller:
            engineID = self.get_engine_id(poller)
```

File: actions/node_create_snmpv3.py
Comment (jjm) [inline actions/node_create_snmpv3.py:89]: Minor comment, In the other code bases I work on the linter would flag this with a warning, as if the log level was above info the substation would still be carried out.

File: actions/node_create_snmpv3.py
Comment (jjm) [reply actions/node_create_snmpv3.py]: Looking at the action config, `poller` defaults to `primary` and these lines 

https://github.com/StackStorm-Exchange/stackstorm-orion/blob/79fa88a33e4b1959187877fde357042396bb2c96/actions/lib/actions.py#L237-L238

Will map `poller` to the correct value. So the if statement and else are not required.

File: actions/node_discover_and_add_interfaces.yaml
Comment (jjm) [inline actions/node_discover_and_add_interfaces.yaml:12]: 👍

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jjm) [inline actions/node_discover_and_add_interfaces_by_name_and_type.py:47]: It would be cleaner here to check if `interface['InterfaceID']` is `0` and then issue the log message and then `continue` onto the next loop.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jjm) [inline actions/node_discover_and_add_interfaces_by_name_and_type.py:45]: This loop has lots of nesting, I'm finding it quite hard to follow what's happening.

File: actions/node_discover_and_add_interfaces_by_name_and_type.yaml
Comment (jjm) [inline actions/node_discover_and_add_interfaces_by_name_and_type.yaml]: Missing final new line.

File: actions/update_interface_properties.py
Comment (jjm) [inline actions/update_interface_properties.py:19]: There looks like some code duplication between this action and the other interface custom action. Can we abstract this away somehow?

File: actions/update_node_custom_properties.py
Comment (jjm) [inline actions/update_node_custom_properties.py:22]: 👍

**Review by jjm — COMMENTED** [review]
Some more comments, apologies last set of comments were during working hours so not as in depth as I should have been.

**Comment by jjm** [comment]
Thanks for the PR @jamesdreid, there's a conflict on `pack.yaml` that would need to be addressed. 

Most of my comments I think are just nice to haves, as @nmaludy is one of the pack maintainers so I'll let him have the final say.

File: actions/lib/actions.py
Comment (jamesdreid) [reply actions/lib/actions.py]: More detailed exception info has been added to code

File: actions/lib/actions.py
Comment (jamesdreid) [reply actions/lib/actions.py]: Agreed. Changed message logic to single line

**Comment by jamesdreid** [comment]
@nmaludy The most recent commits on this PR include an converted, orquesta version of the drain_poller workflow #25.  I have not tested the workflow in my environment (I do not trust the "whitelist" to limit the scope even though it works as expected in the action) but it appears to be correct.   I have also added a fairly simple Action that updates a Node IP address.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jamesdreid) [reply actions/node_discover_and_add_interfaces_by_name_and_type.py:45]: The wish for this Action was to allow us to add interfaces on a device based upon the name of the interface.  The problem in this approach was Orion generates a "Caption" for the interface when you invoke the "DiscoverInterfacesOnNode" verb and the "Caption" data that is generated is not consistent (it depends on if the SNMP includes ifAlias/ifDescr data or not) and it will often include an odd "middot" character so I is difficult to create and manage an input list filter based upon this data.  The ifName data is always consistent and simple to get added to a list, but it is not available to Orion until after the "DiscoverInterfacesOnNode" completes and the interface is added to the Node.  The loop logic here works around that limitation by adding all of the interfaces to the Node that match the types (ifType is returned in the data from the "DiscoverInterfacesOnNode" verb) and Admin Status  (ifAdminStatus is also returned from the "DiscoverInterfacesOnNode" verb) passed as input to the Action.  Once these interfaces have been added to the Node, the ifName data is then retrieved in a subsequent call and all of the interfaces that are returned are compared to the list passed as input to the script.  Any of the interfaces that were added to the Node that are not in the list are then deleted from the Node.

It is a convoluted process but it had to be that way to work around the way Orion executes the "DiscoverInterfacesOnNode" verb, the data that is returned from that execution, and the way it calculates the "Caption" which does not have a direct correlation to SNMP data values.

File: actions/node_discover_and_add_interfaces_by_name_and_type.yaml
Comment (jamesdreid) [reply actions/node_discover_and_add_interfaces_by_name_and_type.yaml]: Checked in now

**Comment by nmaludy** [comment]
@bishopbm1 or @jschoewe you guys have any comments here?

File: actions/node_create_snmpv3.py
Comment (bishopbm1) [reply actions/node_create_snmpv3.py]: I agree this is a lot of duplicated code. Wondering if there is a way we could put this in a generic action and only add the specific stuff in this action and in the node create action.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (bishopbm1) [inline actions/node_discover_and_add_interfaces_by_name_and_type.py:82]: looks like the formatting is off here

**Comment by jamesdreid** [comment]
I will start working on some updates to the README to document the new Actions that have been added along with an updated CHANGELOG to outline all the fixes/updates in this new version.  At this point my branch does not include the API Sensor as I never pulled it in.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (bishopbm1) [inline actions/node_discover_and_add_interfaces_by_name_and_type.py:81]: This above code is pretty similar. Could probably simplify with something like:

```
admin_status = 0
if admin_up_required:
    admin_status = 1
# Loop through values in interface_type list, compare to Discovered interface
# data and add interfaces with IfType values that match what is provided in list
# and have the IfAdminStatus value as "UP" or 1
for iftype in interface_type:
    if interface['ifAdminStatus'] == admin_status and interface['ifType'] == iftype:
        self.logger.info('Interface: {} is of IfType: {}. Adding...'.format(
            interface['Caption'],
            interface['ifType']))
        add_interfaces.append(interface)
```

**Review by bishopbm1 — COMMENTED** [review]
Thanks for the PR! There are a few comments that need to be addressed from myself and @jjm.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jamesdreid) [reply actions/node_discover_and_add_interfaces_by_name_and_type.py:81]: The logic was coded so that if the Admin_Status option was not enabled, it would be ignored completely and both ADMIN UP and ADMIN DOWN interfaces would be added.  This change includes the Admin Status in the check in both cases but in one option will select interfaces that are ADMIN DOWN (0) and another option where they are ADMIN UP.  Since we were adding all of the interfaces of a certain type (usyally type 6/Physical), I wanted to have an option that would exclude any interface that was ADMIN down (which really should be the default for any non-configured interface) but it could be ignored if it needed to be added.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jamesdreid) [reply actions/node_discover_and_add_interfaces_by_name_and_type.py:82]: PyCharm is not complaining about the formatting here.  What exactly is wrong?  PEP8 requires line 78 to be less than 100 characters so 79 and 80 are moved down there to address that.

File: actions/node_create_snmpv3.py
Comment (jamesdreid) [reply actions/node_create_snmpv3.py]: I have updated the poller selection logic as suggested above.  I can look to add the SNMPv3 options to the general Node_Create action but I now have many workflows in place that reference my v3 action by name so it will be a long time and many other edits before I will move be able to support my code a consolidated node create action.

File: actions/node_create_snmpv3.py
Comment (jamesdreid) [reply actions/node_create_snmpv3.py:89]: This was code I pulled in from the Node_Create action and I did not write it, but it appears that the issue with the linter would be related to the use of "format" on the log message.  If that is the case, there are numerous instances in this Action and all the other Orion Pack actions that will trigger the same linter error.

File: actions/node_discover_and_add_interfaces_by_name_and_type.py
Comment (jamesdreid) [reply actions/node_discover_and_add_interfaces_by_name_and_type.py:47]: Seems like all the code that I "borrowed" from the existing actions is problematic. 😃 This was lifted directly from the existing node_discover_and_add_interface action starting on line 42 and will have to be corrected there as well.

File: actions/update_node_custom_properties.py
Comment (jamesdreid) [reply actions/update_node_custom_properties.py:22]: Description updated

**Review by bishopbm1 — APPROVED** [review]
LGTM @nmaludy I'm good with merging this

**Review by nmaludy — APPROVED** [review]

**Review by nmaludy — CHANGES_REQUESTED** [review]
@jamesdreid can you fix the merge conflict and I'll get this merged in

File: pack.yaml
Comment (jamesdreid) [reply pack.yaml]: PR included a second significant update - drain_poller worfkflow converted to Orquesta and was bumped to 0.8.1.  Version now reverted to 0.8.0 as requested.

File: requirements.txt
Comment (jamesdreid) [reply requirements.txt]: Version is pinned to the following: 

> orionsdk>=0.0.3,<1.0.0

**Review by nmaludy — APPROVED** [review]

**Comment by nmaludy** [comment]
@jamesdreid looks like there are still some merge conflicts with master, let me know if you need help resolving them.

**Comment by jamesdreid** [comment]
> 
> 
> @jamesdreid looks like there are still some merge conflicts with master, let me know if you need help resolving them.

Not sure what has to be corrected in order to resolve these conflicts.  There was  a bump to the version in pack.yaml after this PR was submitted, and I have since corrected that and updated my files with a minor version increase to reflect my changes.  The other 2 conflicts in CHANGES and README are descriptions to outline my changes with only my new lines added at the top and some typos corrected.

**Comment by nmaludy** [comment]
@jamesdreid you need to merge `master` into your branch so that the git trees are happy and don't have any conflicts. you probably want to do something like:

```shell
git remote add upstream https://github.com/StackStorm-Exchange/stackstorm-orion.git
git fetch -p --all
git checkout master
git merge upstream/master 
< fix the merge conflicted files > 
< commit and push>
```

**Comment by nmaludy** [comment]
@jamesdreid also see this: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line

**Comment by jamesdreid** [comment]
@nmaludy Looks like those commands fixed it.  Changes from master merged and conflicts resolved.  Thanks for the info on the git fixes as I was thinking it could be corrected with changes in my repo alone.

**Review by nmaludy — APPROVED** [review]

**Comment by nmaludy** [comment]
@jamesdreid thanks for the fixes, merged!

**Pull request merged**
